File: client/src/domain/proposal.ts

// Proposal Types for Veil Transaction Flow

//

export const DEFAULT_WINDOW_BLOCKS = 1008;

export const DEFAULT_REPORT_WINDOW_BLOCKS = 432;

export const DEFAULT_PROPOSAL_EXPIRY_BLOCKS = 144;

export const WINDOW_OPTIONS = [
  { label: '1 day', blocks: 144 },
  { label: '3 days', blocks: 432 },
  { label: '7 days', blocks: 1008 },
  { label: '14 days', blocks: 2016 },
  { label: '30 days', blocks: 4320 },
] as const;

export type B32 = string;

export type Signature = string;

export type PubKey = string;

export type TxCategory = 'Trade' | 'Loan' | 'Service' | 'Other';

export type ReportedOutcome = 'Positive' | 'Negative';

export interface Proposal {
  /** Unique ID derived from hash of proposal data */
  id: B32;

  /** Badge ID of the party creating the proposal */
  proposerBadgeId: B32;

  /** Proposer's badge UTXO (for atomic update) */
  proposerBadgeUtxo: { txid: string; vout: number };

  /** Badge ID of the intended counterparty (optional if using address) */
  counterpartyBadgeId?: B32;

  /** Bitcoin address of counterparty (required if no badge ID) */
  counterpartyAddress?: string;

  /** Transaction value in satoshis */
  value: number;

  /** Category for contextual weighting */
  category: TxCategory;

  /** Settlement window in blocks (time for real-world transaction) */
  windowBlocks: number;

  /** Reporting window in blocks (time to submit outcome after settlement) */
  reportWindowBlocks: number;

  /** Block height when proposal was created */
  createdAt: number;

  /** Block height when proposal expires (if not accepted) */
  expiresAt: number;

  /** Optional memo/description */
  memo?: string;

  /** Proposer's signature over the canonical proposal data */
  proposerSignature: Signature;
}

export interface CreateProposalInput {
  proposerBadgeId: B32;
  proposerBadgeUtxo: { txid: string; vout: number };
  counterpartyBadgeId?: B32;
  counterpartyAddress?: string;
  value: number;
  category: TxCategory;
  windowBlocks?: number;
  reportWindowBlocks?: number;
  memo?: string;
}

export type TransactionStatus =
  | 'pending'    // Proposal created, waiting for counterparty
  | 'active'     // Both signed, settlement window running
  | 'reporting'  // Settlement window closed, awaiting reports
  | 'settled'    // Final state, outcome recorded
  | 'expired'    // Proposal expired before acceptance
  | 'cancelled'; // Proposer withdrew before acceptance

export interface ActiveTransaction {
  /** Unique ID (same as proposal ID) */
  id: B32;

  /** Badge ID of the counterparty */
  counterpartyBadgeId: B32;

  /** Transaction value in satoshis */
  value: number;

  /** Category for contextual weighting */
  category: TxCategory;

  /** Block height when both parties signed (acceptance) */
  startedAt: number;

  /** Block height when settlement window ends (reporting begins) */
  windowEndsAt: number;

  /** Block height when reporting window ends (auto-timeout) */
  reportDeadline: number;

  /** Was I the one who proposed this transaction? */
  iAmProposer: boolean;
}

export interface ReportingTransaction {
  /** Unique ID (same as proposal/active ID) */
  id: B32;

  /** Badge ID of the counterparty */
  counterpartyBadgeId: B32;

  /** Transaction value in satoshis */
  value: number;

  /** Category for contextual weighting */
  category: TxCategory;

  /** Block height when reporting window ends */
  reportDeadline: number;

  /** What I reported (undefined = haven't reported yet) */
  myReport?: ReportedOutcome;

  /** Was I the one who proposed this transaction? */
  iAmProposer: boolean;
}

export type SettledOutcome =
  | { type: 'MutualPositive' }                    // 👍👍
  | { type: 'MutualNegative' }                    // 👎👎
  | { type: 'ContestedIPositive' }                // I said 👍, they said 👎
  | { type: 'ContestedINegative' }                // I said 👎, they said 👍
  | { type: 'Timeout'; reporterOutcome: ReportedOutcome; iReported: boolean }
  | { type: 'MutualTimeout' };                    // Neither reported

export function computeSettledOutcome(
  myReport: ReportedOutcome | undefined,
  theirReport: ReportedOutcome | undefined
): SettledOutcome {
  // Both reported
  if (myReport && theirReport) {
    if (myReport === 'Positive' && theirReport === 'Positive') {
      return { type: 'MutualPositive' };
    }
    if (myReport === 'Negative' && theirReport === 'Negative') {
      return { type: 'MutualNegative' };
    }
    if (myReport === 'Positive' && theirReport === 'Negative') {
      return { type: 'ContestedIPositive' };
    }
    return { type: 'ContestedINegative' };
  }

  // One timed out
  if (myReport && !theirReport) {
    return { type: 'Timeout', reporterOutcome: myReport, iReported: true };
  }
  if (!myReport && theirReport) {
    return { type: 'Timeout', reporterOutcome: theirReport, iReported: false };
  }

  // Both timed out
  return { type: 'MutualTimeout' };
}

export function getOutcomeDescription(outcome: SettledOutcome): string {
  switch (outcome.type) {
    case 'MutualPositive':
      return 'Both parties reported positive';
    case 'MutualNegative':
      return 'Both parties reported negative';
    case 'ContestedIPositive':
      return 'Contested: You reported positive, they reported negative';
    case 'ContestedINegative':
      return 'Contested: You reported negative, they reported positive';
    case 'Timeout':
      return outcome.iReported
        ? `You reported ${outcome.reporterOutcome.toLowerCase()}, they timed out`
        : `They reported ${outcome.reporterOutcome.toLowerCase()}, you timed out`;
    case 'MutualTimeout':
      return 'Neither party reported (treated as positive)';
  }
}

export function getOutcomeTrustWeight(outcome: SettledOutcome): number {
  switch (outcome.type) {
    case 'MutualPositive': return 1.0;
    case 'MutualNegative': return -1.0;
    case 'ContestedIPositive': return 0.0;
    case 'ContestedINegative': return 0.0;
    case 'MutualTimeout': return 0.8;
    case 'Timeout':
      return outcome.reporterOutcome === 'Positive' ? 0.9 : -0.5;
  }
}

export interface OutcomeAggregates {
  /** Both parties said positive 👍👍 */
  mutualPositive: number;

  /** Both parties said negative 👎👎 */
  mutualNegative: number;

  /** I said positive, they said negative */
  contestedIPositive: number;

  /** I said negative, they said positive */
  contestedINegative: number;

  /** One party timed out */
  timeout: number;

  /** Both parties timed out */
  mutualTimeout: number;
}

export function createEmptyAggregates(): OutcomeAggregates {
  return {
    mutualPositive: 0,
    mutualNegative: 0,
    contestedIPositive: 0,
    contestedINegative: 0,
    timeout: 0,
    mutualTimeout: 0,
  };
}

export function getComplaintRatio(agg: OutcomeAggregates): number | null {
  const totalContested = agg.contestedIPositive + agg.contestedINegative;
  if (totalContested === 0) return null;
  return agg.contestedINegative / totalContested;
}

export function getTotalTransactions(agg: OutcomeAggregates): number {
  return (
    agg.mutualPositive +
    agg.mutualNegative +
    agg.contestedIPositive +
    agg.contestedINegative +
    agg.timeout +
    agg.mutualTimeout
  );
}

export function isInSettlementWindow(tx: ActiveTransaction, currentBlock: number): boolean {
  return currentBlock < tx.windowEndsAt;
}

export function isInReportingPhase(tx: ActiveTransaction, currentBlock: number): boolean {
  return currentBlock >= tx.windowEndsAt && currentBlock < tx.reportDeadline;
}

export function isPastDeadline(
  tx: ActiveTransaction | ReportingTransaction,
  currentBlock: number
): boolean {
  return currentBlock >= tx.reportDeadline;
}

export function getBlocksRemaining(targetBlock: number, currentBlock: number): number {
  return Math.max(0, targetBlock - currentBlock);
}

export function blocksToTime(blocks: number): string {
  const minutes = blocks * 10;

  if (minutes < 60) {
    return `${minutes} minutes`;
  }

  const hours = Math.floor(minutes / 60);
  if (hours < 24) {
    return `${hours} hour${hours === 1 ? '' : 's'}`;
  }

  const days = Math.floor(hours / 24);
  return `${days} day${days === 1 ? '' : 's'}`;
}

export function formatSats(sats: number): string {
  if (sats >= 100_000_000) {
    return `${(sats / 100_000_000).toFixed(2)} BTC`;
  }
  if (sats >= 1_000_000) {
    return `${(sats / 1_000_000).toFixed(2)}M sats`;
  }
  if (sats >= 1_000) {
    return `${(sats / 1_000).toFixed(1)}k sats`;
  }
  return `${sats} sats`;
}

export function truncateBadgeId(id: B32, chars: number = 8): string {
  if (id.length <= chars * 2 + 3) return id;
  return `${id.slice(0, chars)}...${id.slice(-chars)}`;
}

File: charms-app/src/types/proposal.rs

use serde::{Deserialize, Serialize};
use serde_with::{serde_as, hex::Hex};
use sha2::{Digest, Sha256};

pub use super::B32;
pub use super::PubKey;
pub use super::Signature;

pub const DEFAULT_WINDOW_BLOCKS: u32 = 1008;

pub const DEFAULT_REPORT_WINDOW_BLOCKS: u32 = 432;

pub const DEFAULT_PROPOSAL_EXPIRY_BLOCKS: u32 = 144;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum TxCategory {
    Trade,
    Loan,
    Service,
    Other,
}

impl Default for TxCategory {
    fn default() -> Self {
        TxCategory::Trade
    }
}

impl TxCategory {
    pub fn as_str(&self) -> &'static str {
        match self {
            TxCategory::Trade => "Trade",
            TxCategory::Loan => "Loan",
            TxCategory::Service => "Service",
            TxCategory::Other => "Other",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "Trade" => Some(TxCategory::Trade),
            "Loan" => Some(TxCategory::Loan),
            "Service" => Some(TxCategory::Service),
            "Other" => Some(TxCategory::Other),
            _ => None,
        }
    }
}

#[serde_as]
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Proposal {
    pub id: B32,

    pub proposer_badge_id: B32,

    pub counterparty_badge_id: B32,

    pub value: u64,

    pub category: TxCategory,

    pub window_blocks: u32,

    pub report_window_blocks: u32,

    pub created_at: u32,

    pub expires_at: u32,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub memo: Option<String>,

    #[serde_as(as = "Hex")]
    pub proposer_signature: Signature,
}

impl Proposal {
    pub fn new(
        proposer_badge_id: B32,
        counterparty_badge_id: B32,
        value: u64,
        category: TxCategory,
        window_blocks: u32,
        report_window_blocks: u32,
        current_block: u32,
        memo: Option<String>,
    ) -> Self {
        let expires_at = current_block.saturating_add(DEFAULT_PROPOSAL_EXPIRY_BLOCKS);

        let mut proposal = Self {
            id: B32::zero(),
            proposer_badge_id,
            counterparty_badge_id,
            value,
            category,
            window_blocks,
            report_window_blocks,
            created_at: current_block,
            expires_at,
            memo,
            proposer_signature: [0u8; 64],
        };

        proposal.id = proposal.compute_id();
        proposal
    }

    pub fn canonical_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend_from_slice(self.proposer_badge_id.as_bytes());
        bytes.extend_from_slice(self.counterparty_badge_id.as_bytes());
        bytes.extend_from_slice(&self.value.to_le_bytes());
        bytes.push(self.category as u8);
        bytes.extend_from_slice(&self.window_blocks.to_le_bytes());
        bytes.extend_from_slice(&self.report_window_blocks.to_le_bytes());
        bytes.extend_from_slice(&self.created_at.to_le_bytes());
        bytes.extend_from_slice(&self.expires_at.to_le_bytes());
        if let Some(ref memo) = self.memo {
            bytes.extend_from_slice(memo.as_bytes());
        }
        bytes
    }

    pub fn compute_id(&self) -> B32 {
        let mut hasher = Sha256::new();
        hasher.update(&self.canonical_bytes());
        let result = hasher.finalize();
        let mut id = [0u8; 32];
        id.copy_from_slice(&result);
        B32(id)
    }

    pub fn signing_hash(&self) -> B32 {
        self.compute_id()
    }

    pub fn is_expired(&self, current_block: u32) -> bool {
        current_block > self.expires_at
    }

    pub fn validate_structure(&self) -> Result<(), ProposalError> {
        if self.value == 0 {
            return Err(ProposalError::ZeroValue);
        }
        if self.window_blocks == 0 {
            return Err(ProposalError::ZeroWindow);
        }
        if self.report_window_blocks == 0 {
            return Err(ProposalError::ZeroReportWindow);
        }
        if self.proposer_badge_id == self.counterparty_badge_id {
            return Err(ProposalError::SelfTransaction);
        }
        if self.id != self.compute_id() {
            return Err(ProposalError::InvalidId);
        }
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ProposalError {
    ZeroValue,
    ZeroWindow,
    ZeroReportWindow,
    SelfTransaction,
    InvalidId,
    Expired,
    InvalidSignature,
    WrongCounterparty,
}

impl core::fmt::Display for ProposalError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::ZeroValue => write!(f, "Transaction value cannot be zero"),
            Self::ZeroWindow => write!(f, "Settlement window cannot be zero"),
            Self::ZeroReportWindow => write!(f, "Report window cannot be zero"),
            Self::SelfTransaction => write!(f, "Cannot transact with yourself"),
            Self::InvalidId => write!(f, "Proposal ID does not match computed hash"),
            Self::Expired => write!(f, "Proposal has expired"),
            Self::InvalidSignature => write!(f, "Invalid proposer signature"),
            Self::WrongCounterparty => write!(f, "You are not the intended counterparty"),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ActiveTransaction {
    pub id: B32,

    pub counterparty_badge_id: B32,

    pub value: u64,

    pub category: TxCategory,

    pub started_at: u32,

    pub window_ends_at: u32,

    pub report_deadline: u32,
}

impl ActiveTransaction {
    pub fn from_proposal(proposal: &Proposal, current_block: u32, i_am_proposer: bool) -> Self {
        let window_ends_at = current_block + proposal.window_blocks;
        let report_deadline = window_ends_at + proposal.report_window_blocks;

        Self {
            id: proposal.id,
            counterparty_badge_id: if i_am_proposer {
                proposal.counterparty_badge_id
            } else {
                proposal.proposer_badge_id
            },
            value: proposal.value,
            category: proposal.category,
            started_at: current_block,
            window_ends_at,
            report_deadline,
        }
    }

    pub fn is_in_settlement_window(&self, current_block: u32) -> bool {
        current_block < self.window_ends_at
    }

    pub fn is_in_reporting_phase(&self, current_block: u32) -> bool {
        current_block >= self.window_ends_at && current_block < self.report_deadline
    }

    pub fn is_past_deadline(&self, current_block: u32) -> bool {
        current_block >= self.report_deadline
    }

    pub fn to_reporting(&self) -> ReportingTransaction {
        ReportingTransaction {
            id: self.id,
            counterparty_badge_id: self.counterparty_badge_id,
            value: self.value,
            category: self.category,
            report_deadline: self.report_deadline,
            my_report: None,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReportingTransaction {
    pub id: B32,

    pub counterparty_badge_id: B32,

    pub value: u64,

    pub category: TxCategory,

    pub report_deadline: u32,

    pub my_report: Option<ReportedOutcome>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum ReportedOutcome {
    Positive,
    Negative,
}

impl ReportingTransaction {
    pub fn has_reported(&self) -> bool {
        self.my_report.is_some()
    }

    pub fn is_past_deadline(&self, current_block: u32) -> bool {
        current_block >= self.report_deadline
    }

    pub fn submit_report(&mut self, outcome: ReportedOutcome) -> Result<(), &'static str> {
        if self.my_report.is_some() {
            return Err("Already reported");
        }
        self.my_report = Some(outcome);
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum SettledOutcome {
    MutualPositive,

    MutualNegative,

    ContestedIPositive,

    ContestedINegative,

    Timeout {
        reporter_outcome: ReportedOutcome,
        i_reported: bool,
    },

    MutualTimeout,
}

impl SettledOutcome {
    pub fn from_reports(
        my_report: Option<ReportedOutcome>,
        their_report: Option<ReportedOutcome>,
    ) -> Self {
        match (my_report, their_report) {

            (Some(ReportedOutcome::Positive), Some(ReportedOutcome::Positive)) => {
                SettledOutcome::MutualPositive
            }
            (Some(ReportedOutcome::Negative), Some(ReportedOutcome::Negative)) => {
                SettledOutcome::MutualNegative
            }
            (Some(ReportedOutcome::Positive), Some(ReportedOutcome::Negative)) => {
                SettledOutcome::ContestedIPositive
            }
            (Some(ReportedOutcome::Negative), Some(ReportedOutcome::Positive)) => {
                SettledOutcome::ContestedINegative
            }

            (Some(outcome), None) => SettledOutcome::Timeout {
                reporter_outcome: outcome,
                i_reported: true,
            },
            (None, Some(outcome)) => SettledOutcome::Timeout {
                reporter_outcome: outcome,
                i_reported: false,
            },

            (None, None) => SettledOutcome::MutualTimeout,
        }
    }

    pub fn is_positive_for_me(&self) -> bool {
        match self {
            SettledOutcome::MutualPositive => true,
            SettledOutcome::MutualTimeout => true,
            SettledOutcome::Timeout { reporter_outcome: ReportedOutcome::Positive, .. } => true,
            _ => false,
        }
    }

    pub fn trust_weight(&self) -> f64 {
        match self {
            SettledOutcome::MutualPositive => 1.0,
            SettledOutcome::MutualNegative => -1.0,
            SettledOutcome::ContestedIPositive => 0.0,
            SettledOutcome::ContestedINegative => 0.0,
            SettledOutcome::MutualTimeout => 0.8,
            SettledOutcome::Timeout { reporter_outcome, .. } => {
                match reporter_outcome {
                    ReportedOutcome::Positive => 0.9,
                    ReportedOutcome::Negative => -0.5,
                }
            }
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct OutcomeAggregates {
    pub mutual_positive: u32,

    pub mutual_negative: u32,

    pub contested_i_positive: u32,

    pub contested_i_negative: u32,

    pub timeout: u32,

    pub mutual_timeout: u32,
}

impl OutcomeAggregates {
    pub fn total(&self) -> u32 {
        self.mutual_positive
            + self.mutual_negative
            + self.contested_i_positive
            + self.contested_i_negative
            + self.timeout
            + self.mutual_timeout
    }

    pub fn total_contested(&self) -> u32 {
        self.contested_i_positive + self.contested_i_negative
    }

    pub fn complaint_ratio(&self) -> Option<f64> {
        let total = self.total_contested();
        if total == 0 {
            return None;
        }
        Some(self.contested_i_negative as f64 / total as f64)
    }

    pub fn record(&mut self, outcome: &SettledOutcome) {
        match outcome {
            SettledOutcome::MutualPositive => self.mutual_positive += 1,
            SettledOutcome::MutualNegative => self.mutual_negative += 1,
            SettledOutcome::ContestedIPositive => self.contested_i_positive += 1,
            SettledOutcome::ContestedINegative => self.contested_i_negative += 1,
            SettledOutcome::Timeout { .. } => self.timeout += 1,
            SettledOutcome::MutualTimeout => self.mutual_timeout += 1,
        }
    }
}
